src/TLuaInterpreter.cpp

```diff
@@ -10483,81 +10526,79 @@ int TLuaInterpreter::echoPopup(lua_State* L)
 
-    if (!lua_isstring(L, s)) {
-        lua_pushstring(L, "echoLink: wrong argument type");
-        lua_error(L);
-        return 1;
+    if (!lua_isstring(L, ++s)) {
+            lua_pushfstring(L, "echoLink: bad argument #%d type (string expected, got %s!)", s, luaL_typename(L, s));
+            return lua_error(L);
     } else {
-        a1 = lua_tostring(L, s);
-        s++;
+        a1 = QString::fromUtf8(lua_tostring(L, s));
     }
     if (n > 1) {
-        if (!lua_isstring(L, s)) {
-            lua_pushstring(L, "echoLink: wrong argument type");
+        if (!lua_isstring(L, ++s)) {
+            lua_pushfstring(L, "echoLink: bad argument #%d type (string expected, got %s!)", s, luaL_typename(L, s));
             lua_error(L);
             return 1;
         } else {
-            a2 = lua_tostring(L, s);
-            s++;
+            a2 = QString::fromUtf8(lua_tostring(L, s));
         }
     }
     if (n > 2) {
-        if (!lua_isstring(L, s)) {
-            lua_pushstring(L, "echoLink: wrong argument type");
+        if (!lua_isstring(L, ++s)) {
+            lua_pushfstring(L, "echoLink: bad argument #%d type (string expected, got %s!)", s, luaL_typename(L, s));
             lua_error(L);
             return 1;
         } else {
-            a3 = lua_tostring(L, s);
-            s++;
+            a3 = QString::fromUtf8(lua_tostring(L, s));
         }
     }
     if (n > 3) {
-        if (lua_isstring(L, s)) {
-            a4 = lua_tostring(L, s);
-            s++;
+        if (lua_isstring(L, ++s)) {
+            a4 = QString::fromUtf8(lua_tostring(L, s));
         } else if (lua_isboolean(L, s)) {
             gotBool = true;
-            a5 = lua_toboolean(L, s);
-            s++;
+            useCurrentFormatElseDefault = lua_toboolean(L, s);
+        } else {
+            lua_pushfstring(L, "echoLink: bad argument #%d type (optionally useCurrentFormatElseDefault boolean or hint as string, got %s!)", s, luaL_typename(L, s));
+            return lua_error(L);
         }
     }
     if (n > 4) {
-        if (!lua_isboolean(L, s)) {
-            lua_pushstring(L, "echoLink: wrong argument type");
-            lua_error(L);
-            return 1;
+        if (!lua_isboolean(L, ++s)) {
+            lua_pushfstring(L, "echoLink: bad argument #%d type (optionally useCurrentFormatElseDefault as boolean expected, got %s!)", s, luaL_typename(L, s));
+            return lua_error(L);
         } else {
-            a5 = lua_toboolean(L, s);
+            useCurrentFormatElseDefault = lua_toboolean(L, s);
             gotBool = true;
-            s++;
         }
     }
+
     Host& host = getHostFromLua(L);
-    QString txt;
-    QString name;
-    QStringList func;
+    QString text;
+    QString windowName;
+    QStringList function;
     QStringList hint;
+
     if (n == 3 || (n == 4 && gotBool)) {
-        txt = a1.c_str();
-        func << a2.c_str();
-        hint << a3.c_str();
-        host.mpConsole->echoLink(txt, func, hint, a5);
```

Comment:
:thinking: Humm, you are not calling this method any more!

Has there been a change already to route all the uses of `echoLink` via a single route (via the `mudlet` class)?  I am planning/in progress in going the other route to send all calls via the "main" `TConsole` instance accessible via the `(TConsole*) Host::mpConsole` member for a number of Lua functions that interact with the `TConsole`s (or `TLabel`s) of the profile (a.k.a. `Host`)...

Reply:
so host.mpConsole is always the 'main' console?

Reply:
Yeah. To be honest I think there is enough extra stuff done in the main `TConsole` instance that it really ought to be made into a separate `TMainConsole` class that inherits the common methods and members of the `TConsole` and has all the extra bits (like the `QMap` containing the names and pointers to all the other sub-consoles) that only it uses moved into it.  As things stand it means that every console has a set of things that serve no purpose in them - like a `TCommandLine` and an emergency stop button etc...

But that is something for another ~day~ ~week~ month?